### ISSUE TITLE
Consider linking Mio's net types w/ an EventLoop

### ISSUE THREAD

**Issue by carllerche** [carllerche/mio#308]
**tl;dr** When a Mio network type is first registered with an `EventLoop` instance, permanently associate the socket with the `EventLoop`.

## Context

On *nix systems, `EventLoop` instances and FDs are fully decoupled. The same FD can be watched from any number of `EventLoop` instances on any number of threads. However, this isn't true on Windows.

On windows, IOCP handles get associated with threads and sockets get associated with IOCP handles. This means that, in order to get the full *nix behavior, a global IOCP handle has to be created and shared across all `EventLoop` instances.

## Proposal

Mio already provides it's own cross platform network types. I propose permanently linking each value with the first `EventLoop` that it gets registered with. This will follow the windows semantics. If an already linked network type is registered with a different `EventLoop`, the register call will return with an `Error`.

If a user of Mio wants to register the same FD with multiple `EventLoop` instances on *nix platforms, this is possible by registering with `EventedFd`.

**Comment by dwrensha** [comment]
Does this change make it now feasible for evented types to deregister themselves when they are dropped?

Today I learned that closing a file descriptor that has been added to `epoll()` does not in fact always remove it from the epoll instance. If the descriptor has been `dup()`ed, as could happen through `try_clone()`, then it will still receive notifications.

This came up in https://github.com/dwrensha/gj/issues/3.

**Comment by carllerche** [comment]
This could make it possible for `mio::TcpStream`, `mio::TcpListener`, and `mio::UdpSocket` ... but more investigation will be required. I'm not 100% sure what the best strategy for this will be. I will come back to this ticket.
